source/meshworker/scratch_data.cc

```diff
@@ -0,0 +1,356 @@
+        return *fe_subface_values;
+      }
+    else
+      return reinit(cell, face_no);
+  }
+
+
+
+  template <int dim, int spacedim>
+  const FEValues<dim, spacedim> &
+  ScratchData<dim, spacedim>::reinit_neighbor(
+    const typename DoFHandler<dim, spacedim>::active_cell_iterator &cell)
+  {
+    if (!neighbor_fe_values)
+      neighbor_fe_values = std::make_unique<FEValues<dim, spacedim>>(
+        *mapping, *fe, cell_quad, neighbor_cell_flags);
+
+    neighbor_fe_values->reinit(cell);
+    cell->get_dof_indices(neighbor_dof_indices);
+    current_neighbor_fe_values = neighbor_fe_values.get();
+    return *neighbor_fe_values;
+  }
+
+
+
+  template <int dim, int spacedim>
+  const FEFaceValues<dim, spacedim> &
+  ScratchData<dim, spacedim>::reinit_neighbor(
+    const typename DoFHandler<dim, spacedim>::active_cell_iterator &cell,
+    const unsigned int                                              face_no)
+  {
+    if (!neighbor_fe_face_values)
+      neighbor_fe_face_values = std::make_unique<FEFaceValues<dim, spacedim>>(
+        *mapping, *fe, face_quad, neighbor_face_flags);
+    neighbor_fe_face_values->reinit(cell, face_no);
+    cell->get_dof_indices(neighbor_dof_indices);
+    current_neighbor_fe_values = neighbor_fe_face_values.get();
+    return *neighbor_fe_face_values;
+  }
+
+
+
+  template <int dim, int spacedim>
+  const FEFaceValuesBase<dim, spacedim> &
+  ScratchData<dim, spacedim>::reinit_neighbor(
+    const typename DoFHandler<dim, spacedim>::active_cell_iterator &cell,
+    const unsigned int                                              face_no,
+    const unsigned int                                              subface_no)
+  {
+    if (subface_no != numbers::invalid_unsigned_int)
+      {
+        if (!neighbor_fe_subface_values)
+          neighbor_fe_subface_values =
+            std::make_unique<FESubfaceValues<dim, spacedim>>(
+              *mapping, *fe, face_quad, neighbor_face_flags);
+        neighbor_fe_subface_values->reinit(cell, face_no, subface_no);
+        cell->get_dof_indices(neighbor_dof_indices);
+        current_neighbor_fe_values = neighbor_fe_subface_values.get();
+        return *neighbor_fe_subface_values;
+      }
+    else
+      return reinit_neighbor(cell, face_no);
+  }
+
+
+
+  template <int dim, int spacedim>
+  const FEValuesBase<dim, spacedim> &
+  ScratchData<dim, spacedim>::get_current_fe_values() const
+  {
+    Assert(current_fe_values != nullptr,
+           ExcMessage("You have to initialize the cache using one of the "
+                      "reinit function first!"));
+    return *current_fe_values;
+  }
+
+
+
+  template <int dim, int spacedim>
+  const FEValuesBase<dim, spacedim> &
+  ScratchData<dim, spacedim>::get_current_neighbor_fe_values() const
+  {
+    Assert(current_neighbor_fe_values != nullptr,
+           ExcMessage("You have to initialize the cache using one of the "
+                      "reinit function first!"));
```

Comment:
```suggestion
                      "reinit functions first!"));
```
Same in a bunch of other places.